cmd/crates/soroban-spec-typescript/src/project_template/src/assembled-tx.ts

```diff
@@ -0,0 +1,706 @@
+     holder needs to sign.
+     */
+    sign: (payload: Buffer) => Buffer | Promise<Buffer>,
+    /**
+     * When to set each auth entry to expire. Could be any number of blocks in
+     * the future. Can be supplied as a promise or a raw number. Default:
+     * contract's current `persistent` storage expiration date/ledger
+     * number/block.
+     */
+    expiration: number | Promise<number> = this.getStorageExpiration('persistent')
+  ): Promise<void> => {
+    if (!this.raw) throw new Error('Transaction has not yet been assembled or simulated')
+    const needsNonInvokerSigningBy = await this.needsNonInvokerSigningBy()
+
+    if (!needsNonInvokerSigningBy) throw new NoUnsignedNonInvokerAuthEntriesError('No unsigned non-invoker auth entries; maybe you already signed?')
+    if (needsNonInvokerSigningBy.indexOf(publicKey) === -1) throw new Error('No auth entries for this public key')
+
+    const rawInvokeHostFunctionOp = this.raw
+      .operations[0] as Operation.InvokeHostFunction
+
+    const originalAuthEntries = rawInvokeHostFunctionOp.auth ?? []
+
+    for (const entry of originalAuthEntries) {
+      if (
+        entry.credentials().switch() !==
+        xdr.SorobanCredentialsType.sorobanCredentialsAddress()
+      ) {
+        // if the invoker/source account, then the entry doesn't need explicit
+        // signature, since the tx envelope is already signed by the source
+        // account, so only check for sorobanCredentialsAddress
+        continue
+      }
+      const pk = StrKey.encodeEd25519PublicKey(
+        entry.credentials().address().address().accountId().ed25519()
+      )
+
+      // this auth entry needs to be signed by a different account
+      // (or maybe already was!)
+      if (pk !== publicKey) continue
+
+
+      const preimage = this.preImageFor(entry, await expiration)
+      const payload = hash(preimage.toXDR())
+      const signature = await sign(payload)
+
+      // TODO: do we need this check?
+      if (!Keypair.fromPublicKey(pk).verify(payload, signature)) {
+        throw new Error(
+          `Signature doesn't match payload!\n\n` +
+          `  publicKey: ${pk}\n` +
+          `  signature: ${signature}`
+        )
+      }
+
+      const sigScVal = nativeToScVal(
+        {
+          public_key: StrKey.decodeEd25519PublicKey(pk),
+          signature,
+        },
+        {
+          // force the keys to be interpreted as symbols (expected for
+          // Soroban [contracttype]s)
+          // Pr open to fix this type in the gen'd xdr
+          type: {
+            public_key: ["symbol", null],
+            signature: ["symbol", null],
+          },
+        },
+      )
+
+      // get a reference to the auth entry, then mutate it
+      const addrAuth = entry.credentials().address()
+      addrAuth.signature(xdr.ScVal.scvVec([sigScVal]))
+
+      addrAuth.signatureExpirationLedger(
+        preimage.sorobanAuthorization().signatureExpirationLedger()
+      )
+    }
```

Comment:
This code is nearly identical to the code of [`authorizeEntry`](https://stellar.github.io/js-soroban-client/global.html#authorizeEntry) (see the [source code](https://stellar.github.io/js-soroban-client/js-stellar-base_src_auth.js.html#line97)); can we leverage that here to simplify things?

Reply:
Probably! Thanks. I started with an older example and wasn't sure what the new interfaces were; just wanted to get something working.